Fix various memory leaks detected by coverity

vigoux opened the pull request:
A bunch of fixes for memory leaks reported by coverity.

Complete list is:
- fix(coverity/352829): correctly free memory in f_call,

vigoux committed:
fix(coverity/352829): correctly free memory in f_call

This function was not freeing allocated memory that it owns when calling
functions from lua.

zeertzjq commented on src/nvim/lua/converter.c:
```diff
@@ -388,12 +388,12 @@ bool nlua_pop_typval(lua_State *lstate, typval_T *ret_tv)
 
       cur.tv->v_type = VAR_FUNC;
-      cur.tv->vval.v_string = (char *)vim_strsave(name);
+      cur.tv->vval.v_string = name;
```
I believe this duplication is necessary, otherwise the entry in the function hash table will be freed.

zeertzjq commented on src/nvim/api/private/converter.c:
```diff
@@ -356,7 +356,7 @@ bool object_to_vim(Object obj, typval_T *tv, Error *err)
       (char *)register_cfunc(&nlua_CFunction_func_call, &nlua_CFunction_func_free, state);
     tv->v_type = VAR_FUNC;
-    tv->vval.v_string = xstrdup(name);
+    tv->vval.v_string = name;
```
ditto

vigoux replied:
Given the test failures, this seems to be the case.
So I'll mark this as a false positive ?

vigoux commented:
Fixed all the mentionned issues. Now I know a little more about all this, thanks !

vigoux commented:
Maybe I'll add a bit more instances of leakages detected by coverity, in order to make that PR grow a bit more

vigoux committed:
fix(coverity/348313): don't exit too early in runtime_default

vigoux committed:
fix(coverity/348300): free memory when overiding sing attribute

Nothing prevent the user from doing `:sign define abc culhl=Normal
culhl=Normal` and thus this leads to an obvious memory leak.

vigoux commented:
So there was an initial attempt at helping coverity estimate the leaks with #18648.
The problem here is:
- `register_cfunc` is not _necessarily_ leaking, which is not the case of the model here. The cases where it leaks are tricky and depend on wether the function is refcounted, etc..
- Some event loop functions do not have any model, and they, too, declare leaked memory while there is not.

vigoux commented:
I think that for now this PR is ready. The question of modelling will be done in an other PR.

jamessan commented:
> * `register_cfunc` is not _necessarily_ leaking, which is not the case of the model here. The cases where it leaks are tricky and depend on wether the function is refcounted, etc..

There isn't an _actual_ leak in `register_cfunc`.  It's just that we're not returning the pointer we allocated, so Coverity can't track how it's managed.  Instead, we stuff a member of the allocated object into a "global" hash and later on use the info in the hash to retrieve the pointer and free the memory.

bfredl commented on src/nvim/memory.h:
```diff
@@ -69,4 +69,9 @@ typedef struct {
     (void)(*ptr_); \
   } while (0)
 
+#define XFREE_CLEARN(ptr) \
```
can just use `XFREE_CLEAR` directly. freeing a null-pointer is a no-op.

vigoux replied:
I was wondering that, then I'll just to that.

vigoux commented:
> > * `register_cfunc` is not _necessarily_ leaking, which is not the case of the model here. The cases where it leaks are tricky and depend on wether the function is refcounted, etc..
> 
> There isn't an _actual_ leak in `register_cfunc`. It's just that we're not returning the pointer we allocated, so Coverity can't track how it's managed. Instead, we stuff a member of the allocated object into a "global" hash and later on use the info in the hash to retrieve the pointer and free the memory.

Then it is not properly "leaked" is it ? I think we can mark it with a special attribute (can't remember the name for now), specifying that, although this returned value is allocated, it is stored and the liberation of it will be handled later on ?

justinmk commented on test/functional/ex_cmds/sign_spec.lua:
```diff
@@ -1,5 +1,5 @@
 local helpers = require('test.functional.helpers')(after_each)
-local clear, nvim, eq = helpers.clear, helpers.nvim, helpers.eq
+local clear, nvim, eq, alive = helpers.clear, helpers.nvim, helpers.eq, helpers.assert_alive
```
Renaming things on import, adds entropy and makes code less consistent.

```suggestion
local clear, nvim, eq, assert_alive = helpers.clear, helpers.nvim, helpers.eq, helpers.assert_alive
```

justinmk commented:
> think we can mark it with a special attribute (can't remember the name for now),

    git grep 'coverity\['

I see that `// coverity[leaked_storage]` was already tried in #18648 , but maybe it needs to be at the end of the line?

justinmk reviewed (approved):
Nice fixes!

vigoux commented:
Regarding the `assert_alive`: my problem is that the test does not fail. What I wanted to check is that, when using ASAN, neovim exited with an exit code different from 0.

I can safely reproduce the bug though.

justinmk commented:
> Regarding the `assert_alive`: my problem is that the test does not fail. What I wanted to check is that, when using ASAN, neovim exited with an exit code different from 0.

Maybe we need a new test util function using `:cquit`. There are hints in `test/functional/core/exit_spec.lua|`.
